### PR TITLE
GitHub Notifications

### PR DESCRIPTION
**Pull request by benibenj**
- With the setting ```githubPullRequests.notifications```, a user can turn GitHub Notifications on and off (currently opt in)
- Uses proposed API

### CODE REVIEW

File: src/github/notifications.ts [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+	public async markAllNotificationsAsRead(owner, repo) {
+		const gitHub = this.getGitHub();
+		await gitHub?.octokit.request('PUT /repos/{owner}/{repo}/notifications', {
+			owner: owner,
+			repo: repo
+		});
+	}
+
+	public async getNotificationThread(thredId) {
```
Comment (alexr00) [inline src/github/notifications.ts:241]: nit: typo in `thredId`.

File: src/github/notifications.ts [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+						currentPrNotifications.push(newNotification);
+					}
+				}
+
+			}
+		}));
+
+		this.adaptPRNotifications();
+
+		this._gitHubPrsTree.updateNotificationBadge(this);
```
Comment (alexr00) [inline src/github/notifications.ts:361]: This `updateNotificationBadge` call is circular as it immediately calls back into the `NotificaitonProvider`. Instead, can you use https://github.com/microsoft/vscode-pull-request-github/blob/ed64fa806143ba1a9cef6c593a1eda9feb24e7c3/src/view/prsTreeDataProvider.ts#L32-L33 to get the view and update the badge directly in `NotificationProvider`

**Review by alexr00 — CHANGES_REQUESTED** [review]

**Comment by alexr00** [comment]
Having a separate notification provider makes sense as the notifications are not repo specific. I have a few questions and changes, but it's heading in a good direction!

File: src/github/notifications.ts [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+
+			if (credentialStore.isAuthenticated(AuthProvider['github-enterprise']) && hasEnterpriseUri()) {
+				this._authProvider = AuthProvider['github-enterprise'];
+			}
+		});
+	}
+
+	public getPrIdentifier(pullRequest: PullRequestModel | OctokitResponse<any>['data']): string {
+		if (pullRequest instanceof PullRequestModel) {
+			return `${pullRequest.remote.url}:${pullRequest.number}`;
```
[the remark this answers is not in the document]
Comment (benibenj) [reply src/github/notifications.ts:118]: The remote url should uniquely identify a repo, no matter if it's a fork or whatever. And then the PR number is enough for picking the PR in that repo

File: src/github/notifications.ts [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+		return (this._authProvider !== undefined) ?
+			this._credentialStore.getHub(this._authProvider) :
+			undefined;
+	}
+
+	public async getNotifications() {
+		const gitHub = this.getGitHub();
+		if (gitHub === undefined)
+			return undefined;
+		const { data, headers } = await gitHub.octokit.request('GET /notifications', {});
```
[the remark this answers is not in the document]
Comment (benibenj) [reply src/github/notifications.ts:221]: Your right, I didn't consider it as it is not mentioned in the API documentation, but it is better to use it.

File: package.json [hunk 5]
```diff
@@ -157,6 +158,15 @@
...
 					"description": "When true, show GitHub Pull Requests within the SCM viewlet. Otherwise show a separate view container for them."
 				},
+				"githubPullRequests.notifications": {
+					"type": "string",
+					"enum": [
+						"pullRequests",
+						"off"
+					],
+					"default": "off",
+					"description": "If Github notifications should be shown to the user."
```
Comment (alexr00) [inline package.json:168]: nit: uppercase H in GitHub.

File: src/github/notifications.ts [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+	private _authProvider: AuthProvider | undefined;
+	// The key uniquely identifies a PR from a Repository. The key is created with `getPrIdentifier`
+	private _notifications: Map<string, Notification[]>;
+	private readonly _reposManager: RepositoriesManager;
+
+	private _pollingDuration: number;
+	private _lastModified: string;
+	private _pollingHandler: NodeJS.Timeout | null;
+
+	private displosables: vscode.Disposable[] = [];
```
Comment (alexr00) [inline src/github/notifications.ts:55]: nit: typo in disposables

File: src/view/prNodeDecorationProvider.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { fromPRNodeUri } from '../common/uri';
+import { NotificationProvider } from '../github/notifications';
+
+export class PRNodeDecorationProvider implements vscode.FileDecorationProvider {
+	private _disposables: vscode.Disposable[] = [];
+
+	_onDidChangeFileDecorations: vscode.EventEmitter<vscode.Uri | vscode.Uri[]> = new vscode.EventEmitter<
```
Comment (alexr00) [inline src/view/prNodeDecorationProvider.ts:13]: Can be `private`

File: src/github/notifications.ts [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+			}
+		}
+
+		// Only update if the user has new notifications
+		if (this._lastModified === headers['last-modified']) {
+			return;
+		}
+		this._lastModified = headers['last-modified'];
+
+		const PrNodesToUpdate = this.uriFromNotifications();
```
Comment (alexr00) [inline src/github/notifications.ts:299]: nit: start variables with lower case.

**Review by alexr00 — CHANGES_REQUESTED** [review]
This looks great, just a few minor comments. Let's merge it after endgame.

File: src/github/notifications.ts (hunk 8 above)
Comment (benibenj) [reply src/github/notifications.ts]: a test comment

**Review by alexr00 — APPROVED** [review]
🚀

**Pull request merged**